File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) 
    {
        int row = matrix.size();
        int col = matrix[0].size();

        vector<int> arr1;
        vector<int> arr2;

        // ✅ Proper 2D traversal
        for(int i = 0; i < row; i++)
        {
            for(int j = 0; j < col; j++)
            {
                if(matrix[i][j] == 0)
                {
                    arr1.push_back(i);
                    arr2.push_back(j);
                }
            }
        }

        int size = arr1.size();

        // ✅ Zero out rows
        for(int i = 0; i < size; i++)
        {
            int rowIdx = arr1[i];
            for(int j = 0; j < col; j++)
            {
                matrix[rowIdx][j] = 0;
            }
        }

        // ✅ Zero out columns
        for(int i = 0; i < size; i++)
        {
            int colIdx = arr2[i];
            for(int j = 0; j < row; j++)
            {
                matrix[j][colIdx] = 0;
            }
        }
    }
};

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.java
class Solution {
  public void setZeroes(int[][] mat) {
    int m = mat.length;
    int n = mat[0].length;

    boolean flagR = false;
    boolean flagC = false;

    // Mark Row
    for(int i = 0; i < m; i++) {
      if(mat[i][0] == 0) flagC = true;
    }

    // Mark Col
    for(int j = 0; j < n; j++) {
      if(mat[0][j] == 0) flagR = true;
    }

    // Use Markers
    for(int i = 1; i < m; i++) {
      for(int j = 1; j < n; j++) {
        if(mat[i][j] == 0) {
            mat[i][0] = 0;
            mat[0][j] = 0;
        }
      }
    }

    // Set Zero Based On Marker
    for(int i = 1; i < m; i++) {
      for(int j = 1; j < n; j++) {
        if(mat[i][0] == 0 || mat[0][j] == 0) {
          mat[i][j] = 0;
        }
      }
    }

    if(flagC) {
      for(int i = 0; i < m; i++) mat[i][0] = 0;
    }

    if(flagR) {
      for(int j = 0; j < n; j++) mat[0][j] = 0;
    }

  }
}